File: go/terminal/ui/controls.go
package ui

import (
	"github.com/nullentropy/caution/go/terminal/gfx"
)

// Progress, slider, radio group and tabs. Local interaction here, semantic
// events upstream, all styling from theme tokens. Ports of src/ui/controls.ts.

var ctrlFont = gfx.NewFont(13, gfx.FontOpts{})

// -- progress -------------------------------------------------------------------

// Progress is a determinate progress bar; Value is 0..1. Display-only.
type Progress struct {
	Core
	Value float32
}

func NewProgress() *Progress { p := &Progress{}; p.self = p; return p }

func (p *Progress) IntrinsicSize() Size { return Size{220, 6} }

func (p *Progress) PaintSelf(dl *gfx.DisplayList) {
	b := p.Bounds
	r := gfx.CornerRadius(b.H / 2)
	dl.Rect(b, *Theme["panelInset"], gfx.RectOpts{Radius: r, BorderWidth: Metrics.BorderWidth, BorderColor: Theme["edgeSoft"]})
	v := min(1, max(0, p.Value))
	if v > 0 {
		dl.Fill(gfx.R(b.X, b.Y, max(b.H, b.W*v), b.H), *Theme["accent"], r)
	}
}

// -- slider ---------------------------------------------------------------------

// Slider is a horizontal slider over [Min, Max], optionally stepped.
// Dragging is client-authoritative (server value sets are ignored mid-drag);
// moves emit input (debounced upstream), release or a key press emits commit.
type Slider struct {
	Core
	Min, Max, Step, Value float32
	OnInput               func(v float32)
	OnCommit              func(v float32)

	Dragging bool
	hovered  bool
}

func NewSlider() *Slider { s := &Slider{Max: 1}; s.self = s; return s }

func (s *Slider) IntrinsicSize() Size {
	return Size{220, max(24, Metrics.SliderThumb*2+4)}
}
func (s *Slider) Interactive() bool { return true }
func (s *Slider) Focusable() bool   { return true }
func (s *Slider) Cursor() string    { return "pointer" }

func (s *Slider) span() float32 {
	if d := s.Max - s.Min; d != 0 {
		return d
	}
	return 1
}

func (s *Slider) ratio() float32 { return min(1, max(0, (s.Value-s.Min)/s.span())) }

func (s *Slider) valueAt(x float32) float32 {
	b := s.Bounds
	thumbR := Metrics.SliderThumb
	usable := max(1, b.W-thumbR*2)
	v := s.Min + min(1, max(0, (x-b.X-thumbR)/usable))*s.span()
	if s.Step > 0 {
		v = s.Min + float32(int((v-s.Min)/s.Step+0.5))*s.Step
	}
	return min(s.Max, max(s.Min, v))
}

func (s *Slider) setValue(v float32, commit bool) {
	if v != s.Value {
		s.Value = v
		s.Invalidate()
		if s.OnInput != nil {
			s.OnInput(v)
		}
	}
	if commit && s.OnCommit != nil {
		s.OnCommit(s.Value)
	}
}

func (s *Slider) OnPointerDown(x, _ float32, _ int) {
	s.Dragging = true
	s.setValue(s.valueAt(x), false)
}

func (s *Slider) OnPointerDrag(x, _ float32) {
	if s.Dragging {
		s.setValue(s.valueAt(x), false)
	}
}

func (s *Slider) OnPointerUp(_, _ float32) {
	if s.Dragging {
		s.Dragging = false
		s.setValue(s.Value, true)
	}
}

func (s *Slider) OnKey(k Key) bool {
	nudge := s.Step
	if nudge <= 0 {
		nudge = s.span() / 100
	}
	var v float32
	switch k.Name {
	case "Left", "Down":
		v = s.Value - nudge
	case "Right", "Up":
		v = s.Value + nudge
	case "Home":
		v = s.Min
	case "End":
		v = s.Max
	default:
		return false
	}
	s.setValue(min(s.Max, max(s.Min, v)), true)
	return true
}

func (s *Slider) OnHoverChange(h bool) {
	s.hovered = h
	s.Invalidate()
}

func (s *Slider) PaintSelf(dl *gfx.DisplayList) {
	b := s.Bounds
	thumbR, trackH := Metrics.SliderThumb, Metrics.SliderTrack
	cy := b.Y + b.H/2
	track := gfx.R(b.X+thumbR, cy-trackH/2, b.W-thumbR*2, trackH)
	dl.Rect(track, *Theme["panelInset"], gfx.RectOpts{
		Radius: gfx.CornerRadius(trackH / 2), BorderWidth: Metrics.BorderWidth, BorderColor: Theme["edgeSoft"],
	})
	tx := track.X + track.W*s.ratio()
	if tx > track.X {
		dl.Fill(gfx.R(track.X, track.Y, tx-track.X, track.H), *Theme["accent"], gfx.CornerRadius(trackH/2))
	}
	thumb := gfx.R(tx-thumbR, cy-thumbR, thumbR*2, thumbR*2)
	if s.UI.ShowFocusRing(s) {
		PaintFocusRing(dl, thumb, thumbR)
	}
	dl.Shadow(thumb, gfx.WithAlpha(gfx.Black, 0.4), 6, gfx.CornerRadius(thumbR), 0, 2)
	fill := *Theme["accent"]
	if s.Dragging {
		fill = gfx.Mix(fill, gfx.Black, 0.15)
	} else if s.hovered {
		fill = gfx.Mix(fill, gfx.White, 0.12)
	}
	white := gfx.White
	dl.Rect(thumb, fill, gfx.RectOpts{Radius: gfx.CornerRadius(thumbR), BorderWidth: Metrics.BorderWidth * 2, BorderColor: &white})
}

// -- radio group ------------------------------------------------------------------

// radioD is the radio circle's diameter: checkbox.size - 2, because a circle
// reads optically larger than a square of the same box.
func radioD() float32 { return Metrics.CheckboxSize - 2 }

// RadioGroup is a vertical radio list; picking emits select with the index.
type RadioGroup struct {
	Core
	Options  []string
	Selected int
	OnSelect func(i int)
	hoverRow int
}

func NewRadioGroup() *RadioGroup {
	r := &RadioGroup{Selected: -1, hoverRow: -1}
	r.self = r
	return r
}

func (r *RadioGroup) IntrinsicSize() Size {
	var w float32
	for _, o := range r.Options {
		w = max(w, ceil32(r.UI.Measure(ctrlFont, o).Width))
	}
	return Size{radioD() + Metrics.SpaceGap + w, float32(len(r.Options)) * Metrics.RowHeight}
}

func (r *RadioGroup) Interactive() bool { return true }
func (r *RadioGroup) Focusable() bool   { return true }
func (r *RadioGroup) Cursor() string    { return "pointer" }

func (r *RadioGroup) pick(i int) {
	if i < 0 || i >= len(r.Options) || i == r.Selected {
		return
	}
	r.Selected = i // local echo
	r.Invalidate()
	if r.OnSelect != nil {
		r.OnSelect(i)
	}
}

func (r *RadioGroup) OnKey(k Key) bool {
	switch k.Name {
	case "Down", "Right":
		r.pick(min(len(r.Options)-1, r.Selected+1))
		return true
	case "Up", "Left":
		r.pick(max(0, r.Selected-1))
		return true
	}
	return false
}

func (r *RadioGroup) OnPointerUp(x, y float32) {
	if Contains(r.Bounds, x, y) {
		r.pick(r.rowAt(y))
	}
}

func (r *RadioGroup) OnPointerHover(_, y float32) {
	if i := r.rowAt(y); i != r.hoverRow {
		r.hoverRow = i
		r.Invalidate()
	}
}

func (r *RadioGroup) OnHoverChange(h bool) {
	if !h && r.hoverRow != -1 {
		r.hoverRow = -1
		r.Invalidate()
	}
}

func (r *RadioGroup) rowAt(y float32) int {
	i := int((y - r.Bounds.Y) / Metrics.RowHeight)
	if i >= 0 && i < len(r.Options) {
		return i
	}
	return -1
}

func (r *RadioGroup) PaintSelf(dl *gfx.DisplayList) {
	b := r.Bounds
	d, rowH := radioD(), Metrics.RowHeight
	// The selected dot keeps a 5px ring of accent around it, floored so it
	// never vanishes under a small checkbox.size.
	dot := max(4, d-10)
	for i, opt := range r.Options {
		y := b.Y + float32(i)*rowH
		cy := y + rowH/2
		circle := gfx.R(b.X, cy-d/2, d, d)
		if i == r.Selected && r.UI.ShowFocusRing(r) {
			PaintFocusRing(dl, circle, d/2)
		}
		if i == r.Selected {
			dl.Fill(circle, *Theme["accent"], gfx.CornerRadius(d/2))
			dl.Fill(gfx.R(circle.X+(d-dot)/2, circle.Y+(d-dot)/2, dot, dot),
				gfx.White, gfx.CornerRadius(dot/2))
		} else {
			edge := *Theme["controlEdge"]
			if i == r.hoverRow {
				edge = gfx.Mix(edge, gfx.White, 0.25)
			}
			dl.Rect(circle, *Theme["panelInset"], gfx.RectOpts{
				Radius: gfx.CornerRadius(d / 2), BorderWidth: Metrics.BorderWidth * 1.5, BorderColor: &edge,
			})
		}
		m := r.UI.Measure(ctrlFont, opt)
		ink := Theme["inkDim"]
		if i == r.Selected {
			ink = Theme["ink"]
		}
		dl.Text(opt, b.X+d+Metrics.SpaceGap, cy-(m.Ascent+m.Descent)/2, ctrlFont, *ink)
	}
}

// -- tabs ---------------------------------------------------------------------------

var tabFont = gfx.NewFont(13, gfx.FontOpts{Weight: 600})

// tabPad is a tab's horizontal text padding; tabsH is control.height plus the
// 2px selection underline the bar reserves below the text.
func tabPad() float32 { return Metrics.ControlPadX }
func tabsH() float32  { return Metrics.ControlHeight + 2 }

// Tabs is a tab bar; picking emits select. Content switching is app policy.
type Tabs struct {
	Core
	Options  []string
	Selected int
	OnSelect func(i int)
	hoverTab int
}

func NewTabs() *Tabs { t := &Tabs{hoverTab: -1}; t.self = t; return t }

func (t *Tabs) IntrinsicSize() Size {
	var w float32
	for _, o := range t.Options {
		w += ceil32(t.UI.Measure(tabFont, o).Width) + tabPad()*2
	}
	return Size{w, tabsH()}
}

func (t *Tabs) Interactive() bool { return true }
func (t *Tabs) Focusable() bool   { return true }
func (t *Tabs) Cursor() string    { return "pointer" }

func (t *Tabs) pick(i int) {
	if i < 0 || i >= len(t.Options) || i == t.Selected {
		return
	}
	t.Selected = i // local echo
	t.Invalidate()
	if t.OnSelect != nil {
		t.OnSelect(i)
	}
}

func (t *Tabs) OnKey(k Key) bool {
	switch k.Name {
	case "Right":
		t.pick(min(len(t.Options)-1, t.Selected+1))
		return true
	case "Left":
		t.pick(max(0, t.Selected-1))
		return true
	}
	return false
}

func (t *Tabs) OnPointerUp(x, y float32) {
	if Contains(t.Bounds, x, y) {
		t.pick(t.tabAt(x))
	}
}

func (t *Tabs) OnPointerHover(x, _ float32) {
	if i := t.tabAt(x); i != t.hoverTab {
		t.hoverTab = i
		t.Invalidate()
	}
}

func (t *Tabs) OnHoverChange(h bool) {
	if !h && t.hoverTab != -1 {
		t.hoverTab = -1
		t.Invalidate()
	}
}

func (t *Tabs) tabAt(x float32) int {
	tx := t.Bounds.X
	for i, o := range t.Options {
		w := ceil32(t.UI.Measure(tabFont, o).Width) + tabPad()*2
		if x >= tx && x < tx+w {
			return i
		}
		tx += w
	}
	return -1
}

func (t *Tabs) PaintSelf(dl *gfx.DisplayList) {
	b := t.Bounds
	hair := Metrics.BorderWidth
	dl.Fill(gfx.R(b.X, b.Y+b.H-hair, b.W, hair), *Theme["edgeSoft"], gfx.Corners{})
	tx := b.X
	for i, opt := range t.Options {
		m := t.UI.Measure(tabFont, opt)
		w := ceil32(m.Width) + tabPad()*2
		sel := i == t.Selected
		if sel && t.UI.ShowFocusRing(t) {
			PaintFocusRing(dl, gfx.R(tx+4, b.Y+4, w-8, b.H-8), Metrics.RadiusControl)
		}
		ink := Theme["inkDim"]
		if sel || i == t.hoverTab {
			ink = Theme["ink"]
		}
		dl.Text(opt, tx+tabPad(), b.Y+(b.H-(m.Ascent+m.Descent))/2-1, tabFont, *ink)
		if sel {
			dl.Fill(gfx.R(tx+tabPad()-4, b.Y+b.H-2, w-tabPad()*2+8, 2), *Theme["accent"], gfx.CornerRadius(1))
		}
		tx += w
	}
}

File: src/ui/controls.ts
import { BLACK, mix, WHITE, withAlpha } from '../gfx/color';
import { font } from '../gfx/font';
import { rect } from '../gfx/geom';
import type { DisplayList } from '../gfx/painter';
import { metrics } from './metrics';
import type { SemanticSpec } from './semantics';
import { theme } from './theme';
import { paintFocusRing } from './widgets';
import { contains, Size, Widget } from './widget';

const CTRL_FONT = font(13);

// -- progress -------------------------------------------------------------------

/** determinate progress bar; `value` is 0..1. Display-only, no events */
export class Progress extends Widget {
  value = 0;

  override intrinsicSize(): Size {
    return { w: this.width ?? 220, h: this.height ?? 6 };
  }

  override semantics(): SemanticSpec {
    return { role: 'text', label: `progress: ${Math.round(this.value * 100)}%` };
  }

  override paintSelf(dl: DisplayList): void {
    const b = this.bounds;
    const r = b.h / 2;
    dl.rect(b, { color: theme.panelInset, radius: r, borderWidth: metrics.borderWidth, borderColor: theme.edgeSoft });
    const v = Math.max(0, Math.min(1, this.value));
    if (v > 0) {
      dl.rect(rect(b.x, b.y, Math.max(b.h, b.w * v), b.h), { color: theme.accent, radius: r });
    }
  }
}

// -- slider ---------------------------------------------------------------------

/**
 * Horizontal slider over [min, max], optionally stepped. Dragging is
 * client-authoritative (the server's value sets are ignored mid-drag, like
 * a focused text field). moves emit `input` (debounced upstream) and the
 * release, or a key press, emits `commit`.
 */
export class Slider extends Widget {
  min = 0;
  max = 1;
  step = 0;
  value = 0;
  onInput: ((v: number) => void) | null = null;
  onCommit: ((v: number) => void) | null = null;

  dragging = false;
  private hovered = false;

  override intrinsicSize(): Size {
    return { w: this.width ?? 220, h: this.height ?? Math.max(24, metrics.sliderThumb * 2 + 4) };
  }

  override get interactive(): boolean {
    return true;
  }

  override get focusable(): boolean {
    return true;
  }

  override cursor(): string {
    return 'pointer';
  }

  override semantics(): SemanticSpec {
    return { role: 'text', label: `slider: ${this.value}` };
  }

  private span(): number {
    return this.max - this.min || 1;
  }

  private ratio(): number {
    return Math.max(0, Math.min(1, (this.value - this.min) / this.span()));
  }

  private valueAt(x: number): number {
    const b = this.bounds;
    const usable = Math.max(1, b.w - metrics.sliderThumb * 2);
    let v = this.min + Math.max(0, Math.min(1, (x - b.x - metrics.sliderThumb) / usable)) * this.span();
    if (this.step > 0) v = this.min + Math.round((v - this.min) / this.step) * this.step;
    return Math.max(this.min, Math.min(this.max, v));
  }

  private setValue(v: number, commit: boolean): void {
    if (v !== this.value) {
      this.value = v;
      this.invalidate();
      this.onInput?.(v);
    }
    if (commit) this.onCommit?.(this.value);
  }

  override onPointerDown(x: number): void {
    this.dragging = true;
    this.setValue(this.valueAt(x), false);
  }

  override onPointerDrag(x: number): void {
    if (this.dragging) this.setValue(this.valueAt(x), false);
  }

  override onPointerUp(): void {
    if (this.dragging) {
      this.dragging = false;
      this.setValue(this.value, true);
    }
  }

  override onKey(e: KeyboardEvent): boolean {
    const nudge = this.step > 0 ? this.step : this.span() / 100;
    let v: number | null = null;
    if (e.key === 'ArrowLeft' || e.key === 'ArrowDown') v = this.value - nudge;
    else if (e.key === 'ArrowRight' || e.key === 'ArrowUp') v = this.value + nudge;
    else if (e.key === 'Home') v = this.min;
    else if (e.key === 'End') v = this.max;
    if (v == null) return false;
    this.setValue(Math.max(this.min, Math.min(this.max, v)), true);
    return true;
  }

  override onHoverChange(hovered: boolean): void {
    this.hovered = hovered;
    this.invalidate();
  }

  override paintSelf(dl: DisplayList): void {
    const b = this.bounds;
    const cy = b.y + b.h / 2;
    const track = rect(b.x + metrics.sliderThumb, cy - metrics.sliderTrack / 2, b.w - metrics.sliderThumb * 2, metrics.sliderTrack);
    dl.rect(track, { color: theme.panelInset, radius: metrics.sliderTrack / 2, borderWidth: metrics.borderWidth, borderColor: theme.edgeSoft });
    const tx = track.x + track.w * this.ratio();
    if (tx > track.x) {
      dl.rect(rect(track.x, track.y, tx - track.x, track.h), { color: theme.accent, radius: metrics.sliderTrack / 2 });
    }
    const thumb = rect(tx - metrics.sliderThumb, cy - metrics.sliderThumb, metrics.sliderThumb * 2, metrics.sliderThumb * 2);
    if (this.ui?.showFocusRing(this)) paintFocusRing(dl, thumb, metrics.sliderThumb);
    dl.shadow(thumb, { blur: 6, color: withAlpha(BLACK, 0.4), offset: [0, 2], radius: metrics.sliderThumb });
    const fill = this.dragging ? mix(theme.accent, BLACK, 0.15) : this.hovered ? mix(theme.accent, WHITE, 0.12) : theme.accent;
    dl.rect(thumb, { color: fill, radius: metrics.sliderThumb, borderWidth: metrics.borderWidth * 2, borderColor: WHITE });
  }
}

// -- radio group ------------------------------------------------------------------

/**
 * The radio circle's diameter: checkbox.size - 2, because a circle reads
 * optically larger than a square of the same box.
 */
const radioD = (): number => metrics.checkboxSize - 2;

/** vertical radio group. picking emits `select` with the option index */
export class RadioGroup extends Widget {
  options: string[] = [];
  selected = -1;
  onSelect: ((i: number) => void) | null = null;
  private hoverRow = -1;

  override intrinsicSize(): Size {
    let w = 0;
    for (const o of this.options) w = Math.max(w, Math.ceil(this.ui!.measure(CTRL_FONT, o).width));
    return { w: radioD() + metrics.spaceGap + w, h: this.options.length * metrics.rowHeight };
  }

  override get interactive(): boolean {
    return true;
  }

  override get focusable(): boolean {
    return true;
  }

  override cursor(): string {
    return 'pointer';
  }

  override semantics(): SemanticSpec {
    return { role: 'text', label: `radio group: ${this.options[this.selected] ?? 'none'}` };
  }

  private pick(i: number): void {
    if (i < 0 || i >= this.options.length || i === this.selected) return;
    this.selected = i; // local echo
    this.invalidate();
    this.onSelect?.(i);
  }

  override onKey(e: KeyboardEvent): boolean {
    if (e.key === 'ArrowDown' || e.key === 'ArrowRight') {
      this.pick(Math.min(this.options.length - 1, this.selected + 1));
      return true;
    }
    if (e.key === 'ArrowUp' || e.key === 'ArrowLeft') {
      this.pick(Math.max(0, this.selected - 1));
      return true;
    }
    return false;
  }

  override onPointerUp(x: number, y: number): void {
    if (contains(this.bounds, x, y)) this.pick(this.rowAt(y));
  }

  override onPointerHover(_x: number, y: number): void {
    const i = this.rowAt(y);
    if (i !== this.hoverRow) {
      this.hoverRow = i;
      this.invalidate();
    }
  }

  override onHoverChange(hovered: boolean): void {
    if (!hovered && this.hoverRow !== -1) {
      this.hoverRow = -1;
      this.invalidate();
    }
  }

  private rowAt(y: number): number {
    const i = Math.floor((y - this.bounds.y) / metrics.rowHeight);
    return i >= 0 && i < this.options.length ? i : -1;
  }

  override paintSelf(dl: DisplayList): void {
    const b = this.bounds;
    const d = radioD();
    const rowH = metrics.rowHeight;
    // the selected dot keeps a 5px ring of accent around it, floored so it
    // never vanishes under a small checkbox.size
    const dot = Math.max(4, d - 10);
    this.options.forEach((opt, i) => {
      const y = b.y + i * rowH;
      const cy = y + rowH / 2;
      const circle = rect(b.x, cy - d / 2, d, d);
      if (i === this.selected && this.ui?.showFocusRing(this)) paintFocusRing(dl, circle, d / 2);
      if (i === this.selected) {
        dl.rect(circle, { color: theme.accent, radius: d / 2 });
        dl.rect(rect(circle.x + (d - dot) / 2, circle.y + (d - dot) / 2, dot, dot), {
          color: WHITE,
          radius: dot / 2,
        });
      } else {
        const edge = i === this.hoverRow ? mix(theme.controlEdge, WHITE, 0.25) : theme.controlEdge;
        dl.rect(circle, { color: theme.panelInset, radius: d / 2, borderWidth: metrics.borderWidth * 1.5, borderColor: edge });
      }
      const m = this.ui!.measure(CTRL_FONT, opt);
      dl.text(opt, b.x + d + metrics.spaceGap, cy - (m.ascent + m.descent) / 2, CTRL_FONT, i === this.selected ? theme.ink : theme.inkDim);
    });
  }
}

// -- tabs ---------------------------------------------------------------------------

const TAB_FONT = font(13, { weight: 600 });

/**
 * A tab's horizontal text padding, and the bar's height: control.height plus
 * the 2px selection underline the bar reserves below the text.
 */
const tabPad = (): number => metrics.controlPadX;
const tabsH = (): number => metrics.controlHeight + 2;

/** Tab bar. Picking emits `select`. */
export class Tabs extends Widget {
  options: string[] = [];
  selected = 0;
  onSelect: ((i: number) => void) | null = null;
  private hoverTab = -1;

  override intrinsicSize(): Size {
    let w = 0;
    for (const o of this.options) w += Math.ceil(this.ui!.measure(TAB_FONT, o).width) + tabPad() * 2;
    return { w, h: tabsH() };
  }

  override get interactive(): boolean {
    return true;
  }

  override get focusable(): boolean {
    return true;
  }

  override cursor(): string {
    return 'pointer';
  }

  override semantics(): SemanticSpec {
    return { role: 'text', label: `tabs: ${this.options[this.selected] ?? ''}` };
  }

  private pick(i: number): void {
    if (i < 0 || i >= this.options.length || i === this.selected) return;
    this.selected = i; // local echo
    this.invalidate();
    this.onSelect?.(i);
  }

  override onKey(e: KeyboardEvent): boolean {
    if (e.key === 'ArrowRight') {
      this.pick(Math.min(this.options.length - 1, this.selected + 1));
      return true;
    }
    if (e.key === 'ArrowLeft') {
      this.pick(Math.max(0, this.selected - 1));
      return true;
    }
    return false;
  }

  override onPointerUp(x: number, y: number): void {
    if (contains(this.bounds, x, y)) this.pick(this.tabAt(x));
  }

  override onPointerHover(x: number, _y: number): void {
    const i = this.tabAt(x);
    if (i !== this.hoverTab) {
      this.hoverTab = i;
      this.invalidate();
    }
  }

  override onHoverChange(hovered: boolean): void {
    if (!hovered && this.hoverTab !== -1) {
      this.hoverTab = -1;
      this.invalidate();
    }
  }

  private tabAt(x: number): number {
    let tx = this.bounds.x;
    for (let i = 0; i < this.options.length; i++) {
      const w = Math.ceil(this.ui!.measure(TAB_FONT, this.options[i]!).width) + tabPad() * 2;
      if (x >= tx && x < tx + w) return i;
      tx += w;
    }
    return -1;
  }

  override paintSelf(dl: DisplayList): void {
    const b = this.bounds;
    const hair = metrics.borderWidth;
    dl.rect(rect(b.x, b.y + b.h - hair, b.w, hair), { color: theme.edgeSoft });
    let tx = b.x;
    this.options.forEach((opt, i) => {
      const m = this.ui!.measure(TAB_FONT, opt);
      const w = Math.ceil(m.width) + tabPad() * 2;
      const selectedTab = i === this.selected;
      if (selectedTab && this.ui?.showFocusRing(this)) {
        paintFocusRing(dl, rect(tx + 4, b.y + 4, w - 8, b.h - 8), metrics.radiusControl);
      }
      const color = selectedTab ? theme.ink : i === this.hoverTab ? theme.ink : theme.inkDim;
      dl.text(opt, tx + tabPad(), b.y + (b.h - (m.ascent + m.descent)) / 2 - 1, TAB_FONT, color);
      if (selectedTab) {
        dl.rect(rect(tx + tabPad() - 4, b.y + b.h - 2, w - tabPad() * 2 + 8, 2), { color: theme.accent, radius: 1 });
      }
      tx += w;
    });
  }
}
